### PR TITLE
Custom FUnction Registration

### PR DESCRIPTION
**Pull request by deepprakash345**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

### CODE REVIEW

File: ui.frontend/src/utils.js [hunk 1]
```diff
@@ -310,12 +311,28 @@ class Utils {
                     _path,
                     _element: elements[i]
                 });
-                const event = new CustomEvent(Constants.FORM_CONTAINER_INITIALISED, { "detail": formContainer });
+                const event = new CustomEvent(Constants.FORM_CONTAINER_INITIALISED, {"detail": formContainer});
```
Comment (nit23uec) [inline ui.frontend/src/utils.js:314]: spaces trimmed..

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java [hunk 2]
```diff
@@ -250,6 +262,29 @@ public String getDataUrl() {
         }
     }
 
+    @Override
+    @JsonIgnore
+    public String getCustomFunctionUrl() {
+        String customFunctionUrl = "";
+        Resource pageResource = resource.getResourceResolver().resolve(this.getParentPagePath());
+        if (pageResource != null && configurationResourceResolver != null) {
+            Resource configResource = configurationResourceResolver.getResource(pageResource, CUSTOM_FUNCTION_CONFIG_BUCKET_NAME,
```
Comment (nit23uec) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java:271]: have you tested it with an anonymous user?

File: ui.frontend/src/utils.js [hunk 3]
```diff
@@ -310,12 +311,28 @@ class Utils {
...
+        if (url != null && url.trim().length > 0) {
+            const customFunctionModule = await import(/*webpackIgnore: true*/ url);
+            const keys = Object.keys(customFunctionModule);
+            for (let i = 0; i < keys.length; i++) {
+                const name = keys[i];
+                const funcDef = customFunctionModule[keys[i]];
+                if (typeof funcDef === 'function') {
+                    const functions = [];
+                    functions[name] = funcDef;
+                    FunctionRuntime.registerFunctions(functions);
```
Comment (nit23uec) [inline ui.frontend/src/utils.js:330]: this call to register functions should be outside for loop.

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java [hunk 4]
```diff
@@ -285,6 +285,16 @@ default String getDataUrl() {
         return null;
     }
 
+    /**
+     * Returns the url to fetch custom functions ESM
```
Comment (nit23uec) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java:289]: `Returns the url to fetch custom functions bundled as an ESM` might have been better.

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java [hunk 5]
```diff
@@ -250,6 +262,29 @@ public String getDataUrl() {
         }
     }
 
+    @Override
+    @JsonIgnore
+    public String getCustomFunctionUrl() {
+        String customFunctionUrl = "";
```
Comment (nit23uec) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java:268]: are you we using any feature toggles in public code nowadays? cc @rismehta

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java (hunk 4 above)
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java]: made the change in the string

File: ui.frontend/src/utils.js (hunk 3 above)
Comment (deepprakash345) [reply ui.frontend/src/utils.js]: done

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java [hunk 6]
```diff
@@ -67,8 +71,16 @@ public class FormContainerImpl extends AbstractContainerImpl implements FormCont
     private static final String DOR_TEMPLATE_TYPE = "dorTemplateType";
     private static final String FD_SCHEMA_TYPE = "fd:schemaType";
     private static final String FD_SCHEMA_REF = "fd:schemaRef";
+    private static final String FD_CUSTOM_FUNCTION_URL = "fd:customFunctionUrl";
```
[the remark this answers is not in the document]
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java]: removed

File: ui.frontend/src/utils.js (hunk 1 above)
Comment (deepprakash345) [reply ui.frontend/src/utils.js]: removed

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java (hunk 5 above)
Comment (rismehta) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java:267]: Using FT on public code is not recommended. We don't use it yet anywhere, if there is a use-case, the design needs to be re-visited.

**Comment by rismehta** [comment]
Please raise PR along with test cases so that it is easy to understand / readable.

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java [hunk 7]
```diff
@@ -69,6 +73,13 @@ public class FormContainerImpl extends AbstractContainerImpl implements FormCont
     private static final String FD_SCHEMA_REF = "fd:schemaRef";
     public static final String FD_FORM_DATA_ENABLED = "fd:formDataEnabled";
 
+    private static final String CUSTOM_FUNCTION_CONFIG_BUCKET_NAME = "settings/cloudconfigs";
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java:76]: configuration-related logic can be managed independently in a separate utility, making this code more organized and easier to maintain, this way sling model remains focused on its main purpose.

```
 private ConfigurationManager configManager = ConfigurationManager.getInstance();
 configManager.getExternalCustomFunctionUrl(configurationResourceResolver);
``

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java [hunk 8]
```diff
@@ -285,6 +285,16 @@ default String getDataUrl() {
         return null;
     }
 
+    /**
+     * Returns the url to fetch custom functions bundled as an ESM
+     *
+     * @return the url to fetch custom functions bundled as an ESM
+     * @since com.adobe.cq.forms.core.components.models.form 4.0.0
+     */
+    default String getCustomFunctionUrl() {
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java:294]: getCustomFunctionModuleUrl

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html [hunk 9]
```diff
@@ -23,6 +23,7 @@
       data-cmp-context-path="${request.contextPath}"
       data-cmp-page-lang="${container.containingPageLang}"
       data-cmp-path="${resource.path}"
+      data-cmp-custom-functions="${container.customFunctionUrl}"
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html:26]: Please add this to the README in optional attributes with the clear purpose

Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html:26]: data-cmp-custom-functions-module-url

Comment (deepprakash345) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html]: done

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java (hunk 8 above)
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java:294]: done

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java (hunk 7 above)
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java:76]: done, implemented as a slingModel after discussion with @rismehta

**Comment by deepprakash345** [comment]
> Please raise PR along with test cases so that it is easy to understand / readable.

Added test case for the new POJO

File: ui.frontend/src/utils.js [hunk 10]
```diff
@@ -296,6 +296,7 @@ class Utils {
                 const _formJson = await HTTPAPILayer.getFormDefinition(_path, _pageLang);
                 console.debug("fetched model json", _formJson);
                 await this.registerCustomFunctions(_formJson.id);
```
Comment (vdua) [inline ui.frontend/src/utils.js:298]: Are we supporting both clientlib and custom function ? I thought it was either of these two.

File: ui.frontend/src/utils.js [hunk 11]
```diff
@@ -316,6 +317,22 @@ class Utils {
         }
     }
 
+    static async registerCustomFunctionsByUrl(url) {
+        if (url != null && url.trim().length > 0) {
+            const customFunctionModule = await import(/*webpackIgnore: true*/ url);
```
Comment (vdua) [inline ui.frontend/src/utils.js:322]: comment would be needed for why have we added webpackIgnore

File: ui.frontend/src/utils.js [hunk 12]
```diff
@@ -316,6 +317,22 @@ class Utils {
...
+            const keys = Object.keys(customFunctionModule);
+            const functions = [];
+            for (let i = 0; i < keys.length; i++) {
+                const name = keys[i];
+                const funcDef = customFunctionModule[keys[i]];
```
Comment (vdua) [inline ui.frontend/src/utils.js:327]: const funcDef = customFunctionModule[name]

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java (hunk 7 above)
Comment (deepprakash345) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java:74]: removed it from the form container model

File: ui.frontend/src/utils.js [hunk 13]
```diff
@@ -316,6 +317,22 @@ class Utils {
         }
     }
 
+    static async registerCustomFunctionsByUrl(url) {
+        if (url != null && url.trim().length > 0) {
+            const customFunctionModule = await import(/*webpackIgnore: true*/ url);
+            const keys = Object.keys(customFunctionModule);
+            const functions = [];
+            for (let i = 0; i < keys.length; i++) {
```
[the remark this answers is not in the document]
Comment (deepprakash345) [reply ui.frontend/src/utils.js:325]: done

File: ui.frontend/src/utils.js (hunk 11 above)
Comment (deepprakash345) [reply ui.frontend/src/utils.js:322]: done

**Pull request merged**